### ISSUE TITLE
RPC failures should include IP in the exception message

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#16923]
When a node to which RPC is sent is down, we get rpc::closed_error. This is often logged as cause of operation failure. For example:
```
ERROR 2024-01-22 23:53:55,654 [shard 0:strm] raft_topology - drain rpc failed, proceed to fence old writes: std::runtime_error (raft topology: exec_global_command(barrier_and_drain) failed with seastar::rpc::closed_error (connection is closed))
```
The caller often doesn't log which IP we attempted to send to, so we can't tell which node caused the operation to fail.

We can either wrap all callers with try/catch which augment the exception with target IP, which is a never ending effort, or change the messaging layer ('send_message()') to include the IP in the exception. I think we should do the latter. At least for rpc::closed_error.

\cc @gleb-cloudius @kbr-scylla

**Comment by mykaul** [comment]
I had a similar ask here - https://github.com/scylladb/scylladb/issues/16588

**Comment by mykaul** [comment]
@emdroid - please see if it's feasible for 2025.4. Looks like it's going to help debugging.

**Comment by emdroid** [comment]
@AbhinavKumarJha Can you have a look at this one? Should be quite straightforward to do.

**Comment by AbhinavKumarJha** [comment]
>change the messaging layer (send_message()) to include the IP in the exception. I think we should do the latter. At least for rpc::closed_error.

This seems difficult, we don't have an option of adding a custom message in `closed_error`, and I dont think we can change error type.
```
class closed_error : public error {
public:
    closed_error() : error("connection is closed") {}
};
```
Another thing we can do here is adding a info log mentioning this stuff in `send_message()`. This will ease the debugging.

Thoughts?

**Comment by emdroid** [comment]
Yeah ... the `closed_error` is a seastar type, so it wouldn't be so easy to enhance it. For now I think we can enhance the `send_message` methods (also `send_message_timeout`, `send_message_cancellable`, `send_message_timeout_cancellable`). In these we actually have the IP address information (`msg_addr id`) ... except for  `send_message_timeout_cancellable` as I can see (which is actually used by gossip).

One option would be to wrap the `closed_error` in a different exception (but that could cause problems with code that wants to catch it on higher levels ... perhaps we could inherit from `closed_error` just for that purpose though). Or just log extra line for each of these cases (which would be the simplest solution).

@patjed41 , @gleb-cloudius any thoughts?

**Comment by emdroid** [comment]
> ... except for `send_message_timeout_cancellable` as I can see (which is actually used by gossip).

We actually have the inet_addr in that method as well, as we're calling `ms->addr_for_host_id(host_id)` there. So the IP should probably be known in all cases.

**Comment by patjed41** [comment]
I'm not sure if logging something in `send_message*` is a good idea. I don't like that then we will have two different logs for one error and we will have to connect them to understand what happened. The `send_message*` log will tell us the node, and another log will tell us the place in the code (like `drain rpc failed, proceed to fence old writes)`.

I think that if we want to do something with `closed_error`, we should do it in Seastar. Maybe we can add a constructor that takes IP and includes it in the error message. I don't know much about Seastar, though.

I see another approach to this issue. It looks like the main problem are logs in the topology coordinator. We can improve `exec_direct_command_helper`:

```cpp
future<> exec_direct_command_helper(raft::server_id id, uint64_t cmd_index, const raft_topology_cmd& cmd) {
    rtlogger.debug("send {} command with term {} and index {} to {}",
        cmd.cmd, _term, cmd_index, id);
    _topology_cmd_rpc_tracker.active_dst.emplace(id);
    auto _ = seastar::defer([this, id] { _topology_cmd_rpc_tracker.active_dst.erase(id); });

    auto result = _db.get_token_metadata().get_topology().is_me(to_host_id(id)) ?
                co_await _raft_topology_cmd_handler(_term, cmd_index, cmd) :
                co_await ser::storage_service_rpc_verbs::send_raft_topology_cmd(
                        &_messaging, to_host_id(id), id, _term, cmd_index, cmd);
    if (result.status == raft_topology_cmd_result::command_status::fail) {
        co_await coroutine::exception(std::make_exception_ptr(
                std::runtime_error(::format("failed status returned from {}", id))));
    }
};
```

We already append `failed status returned from {}` if the command fails. We can do something similar in the case of an exception throw.

That won't fix the issue for all RPCs, but do we even have other RPCs for which can't identify the node? For example, we have no issue with
```
gossip - Failed to notify node b1196e0d-9f50-4b17-b570-d06859623b0a that I am UP: seastar::rpc::closed_error (connection is closed)
```

**Comment by patjed41** [comment]
Maybe the problem is that we do
```cpp
if (f.failed()) {
    co_await coroutine::return_exception(std::runtime_error(
        ::format("raft topology: exec_global_command({}) failed with {}",
            cmd.cmd, f.get_exception())));
}
```
in `exec_global_command_helper`, not `exec_direct_command_helper`, and we lose the information about the node.

**Comment by patjed41** [comment]
> but do we even have other RPCs for which can't identify the node?

We have.

I would go with this

> One option would be to wrap the closed_error in a different exception (but that could cause problems with code that wants to catch it on higher levels ... perhaps we could inherit from closed_error just for that purpose though).

and inherit from `closed_error`.

I think we should also log host ID and IP if available.

**Comment by gleb-cloudius** [comment]
I do not see why we can't add a constructor that gets custom string in seastar and rethrow new instance with updated error message in `send*` functions.

**Comment by patjed41** [comment]
I think we can. It looks like the best idea.
